### PR TITLE
Convert DiskUpdate events to EventChannels

### PR DESCRIPTION
**Pull request by cemozerr**
## PR Description
Converts disk updates and disk update responses to utilize EventChannels instead of EventBus.

### CODE REVIEW

File: data/beaconrestapi/src/test/java/tech/pegasys/artemis/beaconrestapi/BeaconRestApiWithSwaggerTest.java [hunk 1]
```diff
@@ -28,12 +28,13 @@
...
 import tech.pegasys.artemis.storage.ChainStorageClient;
 import tech.pegasys.artemis.storage.CombinedChainDataClient;
+import tech.pegasys.artemis.storage.api.DiskUpdateChannel;
 import tech.pegasys.artemis.sync.SyncService;
 import tech.pegasys.artemis.util.config.ArtemisConfiguration;
 
 public class BeaconRestApiWithSwaggerTest {
   private final ChainStorageClient storageClient =
-      ChainStorageClient.memoryOnlyClient(new EventBus());
+      ChainStorageClient.memoryOnlyClient(new EventBus(), mock(DiskUpdateChannel.class));
```
Comment (ajsutton) [inline data/beaconrestapi/src/test/java/tech/pegasys/artemis/beaconrestapi/BeaconRestApiWithSwaggerTest.java]: I wonder if we should create a text fixture class that creates a memory only `ChainStorageClient` and then it can automatically create the mock `DiskUpdateChannel`.  We never use the memory only version in production so it's weird that we've polluted our production code with test setup constructors.  Might be a separate PR though.

File: data/recorder/src/main/java/tech/pegasys/artemis/data/recorder/SSZTransitionRecorder.java [hunk 2]
```diff
@@ -42,7 +42,7 @@ public SSZTransitionRecorder(final Path outputDirectory) {
   }
 
   @Subscribe
-  public void onGenesisEvent(final StoreGenesisDiskUpdateEvent genesisEvent) {
+  public void onDiskGenesisUpdate(final DiskGenesisUpdate genesisEvent) {
```
Comment (ajsutton) [inline data/recorder/src/main/java/tech/pegasys/artemis/data/recorder/SSZTransitionRecorder.java]: Not sure `DiskGenesisUpdate` is the right name for this event.  It's probably just `GenesisEvent`

File: networking/eth2/src/testFixtures/java/tech/pegasys/artemis/networking/eth2/Eth2NetworkFactory.java [hunk 3]
```diff
@@ -152,11 +156,16 @@ private NetworkConfig generateConfig() {
     }
 
     private void setDefaults() {
+      if (diskUpdateChannel == null) {
+        final EventChannels eventChannels =
+            EventChannels.createSyncChannels(TEST_EXCEPTION_HANDLER, new NoOpMetricsSystem());
+        diskUpdateChannel = eventChannels.getPublisher(DiskUpdateChannel.class);
+      }
```
Comment (ajsutton) [inline networking/eth2/src/testFixtures/java/tech/pegasys/artemis/networking/eth2/Eth2NetworkFactory.java]: This seems like a complicated way of creating a mock that will always return `new SafeFuture()` that doesn't complete.  Since there's no subscriber, nothing will reply so the future never completes.

File: storage/src/main/java/tech/pegasys/artemis/storage/ChainStorageClient.java [hunk 4]
```diff
@@ -59,32 +60,33 @@
...
 
-  public static ChainStorageClient memoryOnlyClient(final EventBus eventBus) {
-    final ChainStorageClient client =
-        new ChainStorageClient(eventBus, TransactionPrecommit.memoryOnly());
+  public static ChainStorageClient memoryOnlyClient(
+      final EventBus eventBus, final DiskUpdateChannel diskUpdateChannel) {
+    final ChainStorageClient client = new ChainStorageClient(diskUpdateChannel, eventBus);
     eventBus.register(client);
     return client;
   }
```
Comment (ajsutton) [inline storage/src/main/java/tech/pegasys/artemis/storage/ChainStorageClient.java]: Given this was just creating a `TransactionPrecommit.memoryOnly()` maybe it should just create a stub `DiskUpdateChannel` implementation that returns the same `DatabaseUpdateResult.successfulWithNothingPruned()` as the result of `onDiskUpdate` calls.

File: storage/src/main/java/tech/pegasys/artemis/storage/ChainStorageClient.java [hunk 5]
```diff
@@ -103,7 +105,8 @@ public void initializeFromGenesis(final BeaconState genesisState) {
           "Failed to set genesis state: store has already been initialized");
     }
 
-    eventBus.post(new StoreGenesisDiskUpdateEvent(store));
+    diskUpdateChannel.onDiskGenesisUpdate(new DiskGenesisUpdate(store));
```
Comment (ajsutton) [inline storage/src/main/java/tech/pegasys/artemis/storage/ChainStorageClient.java]: We don't need the wrapper `DiskGenesisUpdate` event here - just pass `store` as the argument.

File: storage/src/main/java/tech/pegasys/artemis/storage/ChainStorageServer.java [hunk 6]
```diff
@@ -81,15 +83,18 @@ public void onStoreRequest(final GetStoreRequest request) {
...
   }
 
-  @Subscribe
-  public void onStoreDiskUpdate(final StoreDiskUpdateEvent event) {
-    final DatabaseUpdateResult result = database.update(event);
-    handleStoreUpdate(result);
-    eventBus.post(new StoreDiskUpdateCompleteEvent(event.getTransactionId(), result));
+  @Override
+  public SafeFuture<DiskUpdateResult> onDiskUpdate(final DiskUpdate event) {
+    return SafeFuture.supplyAsync(
```
Comment (ajsutton) [inline storage/src/main/java/tech/pegasys/artemis/storage/ChainStorageServer.java]: We shouldn't use `supplyAsync` here or events will be recorded in parallel and potentially out of order.  Just `SafeFuture.of`.

Comment (ajsutton) [reply storage/src/main/java/tech/pegasys/artemis/storage/ChainStorageServer.java]: Actually now I think about it, as a follow up to this we should be able to remove the `.join()` calls after each `commit()` because we know each change will be applied in order.

Though there is still a question of what to do if the returned future throws an `Exception`. Will need some thought, but would be really nice if we can let our processing run ahead of storage.

File: storage/src/main/java/tech/pegasys/artemis/storage/api/DiskUpdateChannel.java [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+ */
+
+package tech.pegasys.artemis.storage.api;
+
+import tech.pegasys.artemis.storage.events.diskupdates.DiskGenesisUpdate;
+import tech.pegasys.artemis.storage.events.diskupdates.DiskUpdate;
+import tech.pegasys.artemis.storage.events.diskupdates.DiskUpdateResult;
+import tech.pegasys.artemis.util.async.SafeFuture;
+
+public interface DiskUpdateChannel {
```
Comment (ajsutton) [inline storage/src/main/java/tech/pegasys/artemis/storage/api/DiskUpdateChannel.java]: I wonder if we should be using `Storage` instead of `Disk` for these classes.  The storage system doesn't necessarily have to store to disk (it could be S3StorageServer etc).

File: storage/src/test/java/tech/pegasys/artemis/storage/MapDbDatabaseTest.java [hunk 8]
```diff
@@ -118,7 +116,7 @@ public void shouldRecreateOriginalGenesisStore() {
 
   @Test
   public void shouldGetHotBlockByRoot() {
-    final Transaction transaction = store.startTransaction(databaseTransactionPrecommit);
+    final Transaction transaction = store.startTransaction(new StubDiskUpdateChannel());
```
Comment (ajsutton) [inline storage/src/test/java/tech/pegasys/artemis/storage/MapDbDatabaseTest.java]: Slightly odd that we're using `StubDiskUpdateChannel` here but a mock in a lot of other places.  Seems like we could just be using the stub.

**Review by ajsutton — APPROVED** [review]
LGTM.  Left some comments around style - the only really important change is to not use `supplyAsync` so we ensure we're processing storage updates in sequence.

File: data/beaconrestapi/src/test/java/tech/pegasys/artemis/beaconrestapi/BeaconRestApiWithSwaggerTest.java (hunk 1 above)
Comment (cemozerr) [reply data/beaconrestapi/src/test/java/tech/pegasys/artemis/beaconrestapi/BeaconRestApiWithSwaggerTest.java]: I highly agree. I actually want to do it in this PR. ChainStorageClient would be way more legible if cleaned from the non-production code.

File: data/recorder/src/main/java/tech/pegasys/artemis/data/recorder/SSZTransitionRecorder.java (hunk 2 above)
Comment (cemozerr) [reply data/recorder/src/main/java/tech/pegasys/artemis/data/recorder/SSZTransitionRecorder.java]: Done.

File: storage/src/main/java/tech/pegasys/artemis/storage/ChainStorageClient.java (hunk 5 above)
Comment (cemozerr) [reply storage/src/main/java/tech/pegasys/artemis/storage/ChainStorageClient.java]: Done.

File: storage/src/main/java/tech/pegasys/artemis/storage/ChainStorageServer.java (hunk 6 above)
Comment (cemozerr) [reply storage/src/main/java/tech/pegasys/artemis/storage/ChainStorageServer.java]: Makes sense. Done.

Comment (cemozerr) [reply storage/src/main/java/tech/pegasys/artemis/storage/ChainStorageServer.java]: Thinking out loud, maybe, in a scenario where the disk is throwing exceptions, we can start logging warning level messages, but keep executing since being able to validate is more important than writing to disk. Through the messages, the user would know something needs fixing, but still not lose their stake in the meanwhile.

Comment (cemozerr) [reply storage/src/main/java/tech/pegasys/artemis/storage/ChainStorageServer.java]: https://pegasys1.atlassian.net/secure/RapidBoard.jspa?rapidView=67&modal=detail&selectedIssue=BC-319

File: storage/src/main/java/tech/pegasys/artemis/storage/api/DiskUpdateChannel.java (hunk 7 above)
Comment (cemozerr) [reply storage/src/main/java/tech/pegasys/artemis/storage/api/DiskUpdateChannel.java]: Interesting point. I did not think of that. I'll make the change.

Comment (cemozerr) [reply storage/src/main/java/tech/pegasys/artemis/storage/api/DiskUpdateChannel.java]: Done.

File: storage/src/test/java/tech/pegasys/artemis/storage/MapDbDatabaseTest.java (hunk 8 above)
Comment (cemozerr) [reply storage/src/test/java/tech/pegasys/artemis/storage/MapDbDatabaseTest.java]: It's because this is not actually a stub (I named it incorrectly) but an interface that handles writing to actual disk without requiring ChainStorageServer class in the process. I'll change the naming and make this more visible.

Comment (cemozerr) [reply storage/src/test/java/tech/pegasys/artemis/storage/MapDbDatabaseTest.java]: It also writes to the updateResults list in the process, which is a requirement for some of the tests.

Comment (cemozerr) [reply storage/src/test/java/tech/pegasys/artemis/storage/MapDbDatabaseTest.java]: Done.

File: networking/eth2/src/testFixtures/java/tech/pegasys/artemis/networking/eth2/Eth2NetworkFactory.java (hunk 3 above)
Comment (cemozerr) [reply networking/eth2/src/testFixtures/java/tech/pegasys/artemis/networking/eth2/Eth2NetworkFactory.java]: Agreed. Thanks for catching. Changed.

File: data/beaconrestapi/src/test/java/tech/pegasys/artemis/beaconrestapi/BeaconRestApiWithSwaggerTest.java (hunk 1 above)
Comment (cemozerr) [reply data/beaconrestapi/src/test/java/tech/pegasys/artemis/beaconrestapi/BeaconRestApiWithSwaggerTest.java]: Actually, I will merge this in and quickly do a follow-up PR. I have a feeling that, that PR might get a lot of comments 😂
